jaxopt/_src/levenberg_marquardt.py

```diff
@@ -0,0 +1,372 @@
+    local minimum of the cost function F(x):
+
+    ```
+    argmin_x F(x) = 0.5 * sum(f_i(x)**2), i = 0, ..., m - 1
+    f(x) = func(x, *args)
+    ```
+
+    The convergence is achieved once the coeff update satisfies the criteria:
+    ```
+    ||dcoeffs||_2 <= xtol * (||coeffs||_2 + xtol)
+    ```
+    or the gradient satisfies ``||grad(f)||_inf <= gtol``.
+
+  Attributes:
+    residual_fun: a smooth function of the form ``residual_fun(x, *args,
+      **kwargs)``.
+    maxiter: maximum increase_factormber of iterations.
+    damping_parameter: The parameter which adds a correction to the equation
+      derived for updating the coefficients using Gauss-Newton method. Please
+      see section 3.2. of K. Madsen et al. in the book "Methods for nonlinear
+      least squares problems" for more information.
+    stop_criteria: The criteria to use for the convergence of the while loop.
+      e.g., for 'madsen-nielsen' the criteria is to satisfy the two equations
+      for delta_params and gradient that is mentioned above. If 'grad-l2' is
+      selected, the convergence is achieved if l2 of gradient is smaller or
+      equal to tol.
+    tol: tolerance.
+    xtol: float, optional The convergence tolerance for the second norm of the
+      coefficient update.
+    gtol: float, optional The convergence tolerance for the inf norm of the
+      residual gradient.
+    solver: str, optional The solver to use when finding dcoeffs, the update to
+      the coeffs in each iteration. This is done through solving a system of
+      linear equation Ax=b. The options are 'cg' (conjugate gradient),
+      'bicgstab' (biconjugate gradient), 'gmres' (generalized minimum residual),
+      'cholesky' (Cholesky factorization), 'inv' (explicit multiplication with
+      matrix inverse). 'cg', 'bicgstab', and 'gmres' are more scalable for
+      runtime but take longer compilations. 'cholesky' is faster than 'inv'
+      since it uses the symmetry feature of A.
+    geodesic: bool, if we would like to include the geodesic acceleration when
+      solving for the delta_params in every iteration.
+    implicit_diff: bool, whether to enable implicit diff or autodiff of unrolled
+      iterations.
+    implicit_diff_solve: the linear system solver to use.
+    verbose: bool, whether to print error on every iteration or not.
+      Warning: verbose=True will automatically disable jit.
+    jit: whether to JIT-compile the bisection loop (default: "auto").
+    unroll: whether to unroll the bisection loop (default: "auto").
+
+  Reference: This algorithm is for finding the best fit parameters based on the
+    algorithm 6.18 provided by K. Madsen & H. B. Nielsen in the book
+    "Introduction to Optimization and Data Fitting".
+  """
+  residual_fun: Callable
+
+  maxiter: int = 30
+  tol: float = 0.0
+
+  damping_parameter: float = 1e-6
+  stop_criteria: Literal['grad-l2', 'madsen-nielsen'] = 'madsen-nielsen'
+  tol: float = 0.0
+  xtol: float = 1e-8
```

Comment:
`1e-8` may be very difficult to reach for float32 precision, which is the default precision we operate with. How about `1e-3`?